scipy/stats/_unuran/unuran_callback.h

```diff
@@ -0,0 +1,271 @@
+    int flags = CCALLBACK_OBTAIN;
+
+    callback->c_function = NULL;
+    callback->user_data = NULL;
+    callback->signature = NULL;
+
+    unur_callback->callbacks = fcn_dict;
+    Py_INCREF(fcn_dict);
+    unur_callback->params = extra_args;
+    Py_INCREF(extra_args);
+
+    /* The required callbacks are present in the `fcn_dict` dictionary. There can
+       be no callbacks for UNU.RAN yet we would need a C callback for error handler
+       to pass on the captured UNU.RAN errors to Python. Hence, we pack a dummy
+       function in the C callback object and use `unuran_callback_t` to obtain
+       required callbacks, if any. */
+    dummy_fcn = PyRun_String("lambda: None", Py_eval_input, PyEval_GetGlobals(), NULL);
+    if (dummy_fcn == NULL) {
+        goto fail;
+    }
+
+    ret = ccallback_prepare(callback, unuran_call_signatures, dummy_fcn, flags);
+    if (ret == -1) {
+        goto fail;
+    }
+
+    callback->info_p = (void *)unur_callback;
+
+    return 0;
+
+fail:
+    Py_DECREF(fcn_dict);
+    Py_DECREF(extra_args);
+    Py_XDECREF(dummy_fcn);
+    return -1;
+}
+
+/**
+ * Release a UNU.RAN callback.
+ * 
+ * Parameters
+ * ----------
+ * callback : ccallback_t *
+ *     C callback to be released
+ * unur_callback : unuran_callback_t *
+ *     UNU.RAN callback to be released
+ * 
+ * Returns
+ * -------
+ * success : int
+ *     0 if success -1 otherwise
+ */
+int release_unuran_callback(ccallback_t *callback, unuran_callback_t *unur_callback) {
+    Py_XDECREF(unur_callback->callbacks);
+    Py_XDECREF(unur_callback->params);
+    Py_XDECREF(callback->py_function);
+    unur_callback->callbacks = NULL;
+    unur_callback->params = NULL;
+    int ret = ccallback_release(callback);
+    return ret;
+}
+
+
+void error_handler(const char *objid, const char *file, int line, const char *errortype,
+                   int unur_errno, const char *reason)
+{
+    ccallback_t *err_callback;
+    const char *errno_msg;
+    char reason_[256], objid_[256], errno_msg_[256];
+
+    if (unur_errno != UNUR_SUCCESS) {
+        err_callback = ccallback_obtain();
+
+        if (reason == NULL || strcmp(reason, "") == 0) {
+            strcpy(reason_, "unknown error");
+        }
+        else {
+            strcpy(reason_, reason);
+        }
+        if (objid == NULL || strcmp(objid, "") == 0) {
+            strcpy(objid_, "unknown");
+        }
+        else {
+            strcpy(objid_, objid);
+        }
+
+        errno_msg = unur_get_strerror(unur_errno);
+
+        if (errno_msg == NULL || strcmp(errno_msg, "") == 0) {
+            strcpy(errno_msg_, "unknown type");
+        }
+        else {
+            strcpy(errno_msg_, errno_msg);
+        }
+        if (strcmp(errortype, "error") == 0) {
+            PyErr_Format(PyExc_RuntimeError,
+                         "[objid: %s] %d : %s => %s",
+                         objid_, unur_errno, reason_,
+                         errno_msg_);
+            longjmp(err_callback->error_buf, 1);
```

Comment:
This error handler handles errors occurring in UNU.RAN. I set the Python error variable and jump out of the C function to raise the error in Cython. I need a `ccallback_t` object to do that (in a thread-safe manner) irrespective of whether Python callbacks are present or not.